File: app/src/main/java/com/king/frame/base/BaseAdapter.java
package com.king.frame.base;

import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;


/**
 * RecyclerView的Adapter
 * @author king
 * @date 2016/10/10 16:33
 */
public abstract class BaseAdapter<T> extends RecyclerView.Adapter<BaseViewHolder>  {

    protected OnItemClickListener mOnItemClickListener;
    protected OnItemLongClickListener mOnItemLongClickListener;
    protected List<T> mDataList = new ArrayList<T>();

    @Override
    public BaseViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return getViewHolder(parent,viewType);
    }

    @Override
    public void onBindViewHolder(BaseViewHolder holder, final int position) {
       holder.bindViewData(position);
    }


    @Override
    public int getItemCount() {
        if(mDataList==null){
            return 0;
        }
        return mDataList.size();
    }

    public abstract BaseViewHolder getViewHolder(ViewGroup parent, int viewType);


    //item点击事件
    public interface OnItemClickListener{
        void onItemClicked(Object o, int pos);
    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        mOnItemClickListener = onItemClickListener;
    }

    //长按点击事件
    public interface OnItemLongClickListener{
        void onItemClicked(Object o, int pos);
    }

    public void setOnItemLongClickListener(OnItemLongClickListener onItemLongClickListener) {
        mOnItemLongClickListener = onItemLongClickListener;
    }

    public void addAll(List<T> dataList){
        mDataList.addAll(dataList);
        notifyDataSetChanged();
    }

    public void add(T t){
        mDataList.add(t);
        notifyItemInserted(mDataList.size());
    }

    public void remove(int position){
        if(mDataList.size() == 0){
            return;
        }
        mDataList.remove(position);
        notifyItemRemoved(position);
    }

    public void onDataChanged(List<T> dataList){
        mDataList = dataList;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/king/frame/base/BaseAdapter.kt
package com.king.frame.base

import android.view.ViewGroup

import java.util.ArrayList

import androidx.recyclerview.widget.RecyclerView


/**
 * RecyclerView的通用Adapter
 *
 * @author king
 * @date 2019-11-27 14:33
 */
abstract class BaseAdapter<T> : RecyclerView.Adapter<BaseViewHolder>() {

    protected var mOnItemClickListener: OnItemClickListener? = null
    protected var mOnItemLongClickListener: OnItemLongClickListener? = null
    protected var mDataList: MutableList<T>? = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
        return getViewHolder(parent, viewType)
    }

    override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
        holder.bindViewData(position)
    }


    override fun getItemCount(): Int {
        return if (mDataList == null) {
            0
        } else mDataList!!.size
    }

    abstract fun getViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder


    //item点击事件
    interface OnItemClickListener {
        fun onItemClicked(o: Any, pos: Int)
    }

    fun setOnItemClickListener(onItemClickListener: OnItemClickListener) {
        mOnItemClickListener = onItemClickListener
    }

    //长按点击事件
    interface OnItemLongClickListener {
        fun onItemClicked(o: Any, pos: Int)
    }

    fun setOnItemLongClickListener(onItemLongClickListener: OnItemLongClickListener) {
        mOnItemLongClickListener = onItemLongClickListener
    }

    fun addAll(dataList: List<T>) {
        mDataList!!.addAll(dataList)
        notifyDataSetChanged()
    }

    fun add(t: T) {
        mDataList!!.add(t)
        notifyItemInserted(mDataList!!.size)
    }

    fun remove(position: Int) {
        if (mDataList!!.size == 0) {
            return
        }
        mDataList!!.removeAt(position)
        notifyItemRemoved(position)
    }

    fun onDataChanged(dataList: MutableList<T>) {
        mDataList = dataList
        notifyDataSetChanged()
    }
}
